server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java

```diff
@@ -237,12 +241,16 @@ public void collectSearchShards(String routing, IntConsumer consumer) {
     public static class ExtractFromSource extends IndexRouting {
         private final Predicate<String> isRoutingPath;
         private final XContentParserConfiguration parserConfig;
+        private final boolean indexTracksRoutingId;
+
+        private int routingId = 0;
```

Comment:
I think in some cases we reuse this instance and if then  `indexShard(...)` gets invoked by multiple callers then this gets tricky around what routing id will be read in postProcess.

Reply:
Right, switched to use a Consumer.